blockchyp/blockchyp_client.py

```diff
@@ -101,49 +108,85 @@ def is_terminal_routed(self, request):
-                "auth_header": ""
-                } # temporary until CryptoUtils is completed
-            config["Headers"] = {
+        if (self.credentials and self.credentials["api_key"]):
+            crypto = CryptoUtils()
+            headers = crypto.generate_gateway_headers(self.credentials)
+            return {
                 "Nonce": headers["nonce"],
                 "Timestamp": headers["timestamp"],
-                "Authorization": headers["auth_header"]
+                "Authorization": headers["auth_header"],
                 }
 
-        return config
-
-    def __get_terminal_config(self):
-        pass
+        return {}
 
     def __gateway_post(self, path, payload):
         pass
 
     def __terminal_get(self, terminal, path, creds):
         pass
 
-    def __terminal_post(self, route, path, payload):
-        pass
+    def __terminal_post(self, route, path, payload=None):
+        url = self.__assemble_terminal_url(route, path)
+        transient_credentials = route["transientCredentials"]
+        data = {
+            "apiKey": transient_credentials["apiKey"],
+            "bearerToken": transient_credentials["bearerToken"],
+            "signingKey": transient_credentials["signingKey"],
+            "request": payload
+        }
+        print(data)
+        reply = requests.post(url, json=data)
+
+        return reply
 
     def __assemble_terminal_url(self, route, path):
-        pass
+        result = "http"
+        if self.https:
+            result = result + "s"
+        result = result + "://"
+        result = result + route["ipAddress"]
+        if self.https:
+            result = result + ":8443"
+        else:
+            result = result + ":8080"
+        result = result + "/api"
+        result = result + path
+
+        return result
 
     def __resolve_terminal_route(self, terminal_name):
-        pass
+        cache_entry = self._route_cache.get(terminal_name, None)
+
+        if cache_entry:
+            # check cache expiration
+            if cache_entry.ttl > int(datetime.utcnow().timestamp()) * 1000:
+                return cache_entry["route"]
+
+        route_response = self.__gateway_get("/terminal-route?terminal=" + terminal_name)
+        route = route_response.json()
+        time = int(datetime.utcnow().timestamp()) * 1000
+        self._route_cache[terminal_name] = {
+            "ttl": time + self.route_cache_ttl * 60000,
+            "route": route,
+        }
+
+        return route
 
 class BlockChypCredentials:
     """
     Stores the three credential components for easy access
     """
     def __init__(self, api_key, bearer_token, signing_key):
-        self.api_key = api_key
-        self.bearer_token = bearer_token
-        self.signing_key = signing_key
+        self.creds = {
+            "api_key": api_key,
+            "bearer_token": bearer_token,
+            "signing_key": signing_key,
+        }
+
+    def __getitem__(self, item):
+        return self.creds[item]
```

Comment:
In the original implementation, you could access fields with: `credentials.api_key`

In this new implementation, you need `credentials["api_key"]`. Why is this better?